numcodecs/gzip.py

```diff
@@ -38,24 +38,35 @@ def encode(self, buf):
                             mode='wb',
                             compresslevel=self.level) as compressor:
             compressor.write(buf)
-        compressed = compressed.getvalue()
 
-        return compressed
+        try:
+            compressed = compressed.getbuffer()
+        except AttributeError:  # pragma: py3 no cover
+            compressed = compressed.getvalue()
+
+        return ensure_ndarray(compressed)
 
     # noinspection PyMethodMayBeStatic
     def decode(self, buf, out=None):
 
         # normalise inputs
-        buf = ensure_contiguous_ndarray(buf)
-        if out is not None:
-            out = ensure_contiguous_ndarray(out)
+        if PY2:  # pragma: py3 no cover
+            buf = ensure_contiguous_ndarray(buf)
+        else:  # pragma: py2 no cover
+            buf = ensure_bytes(buf)
```

Comment:
I'm surprised about this. Is it really necessary to ensure bytes on PY3? That will imply a copy if buf is not bytes. Previously ``ensure_contiguous_ndarray(buf)`` was working on all Pythons?

Reply:
Basically it [always copies on Python 2]( https://github.com/python/cpython/blob/2.7/Modules/_io/bytesio.c#L798 ), but will [skip copying on Python 3]( https://github.com/python/cpython/blob/3.7/Modules/_io/bytesio.c#L923 ) if it's already a `bytes` object. So the idea here is if we already have a `bytes` object (e.g. reading from all(?) Zarr stores), then we should try to pass the `bytes` object through unchanged. If we coerce a copy in the process on Python 3, this is no worse as `BytesIO` would have copied it anyways if it wasn't a `bytes` object.

Reply:
Ah, got it. Thanks.

Reply:
Maybe worth adding a comment here, just so we don't forget why bytes is preferred.

Reply:
Sure, have added comments to explain why `ensure_contiguous_ndarray` is used in the Python 2 case and why `ensure_bytes` is used in the Python 3 case. HTH